coding_systems/base/import_data_utils.py

```diff
@@ -0,0 +1,121 @@
+        self.coding_system = coding_system
+        self.release_name = release_name
+        self.valid_from = valid_from
+        self.import_ref = import_ref or ""
+        self.import_datetime = timezone.now()
+        self.cs_release = None
+        self.new_db_path = None
+        self.backup_path = None
+
+    def __enter__(self):
+        # return the database_alias after setting up the new import db
+        try:
+            self.setup_new_import_database()
+            print(f"Importing {self.coding_system} data...")
+            return self.cs_release.database_alias
+        except Exception:
+            # if anything went wrong during the setup, ensure that any paths have been reset
+            if self.backup_path is not None:
+                shutil.copy(self.backup_path, self.new_db_path)
+                self.backup_path.unlink()
+            elif self.new_db_path is not None and self.new_db_path.exists():
+                self.new_db_path.unlink()
+            raise
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        if exc_value:
+            # If anything went wrong, revert to the initial state
+            if self.backup_path is not None:
+                # there was an existing database (i.e. this import was a forced-overwrite of an
+                # existing one).  Revert the database to the backup, and leave the CodingSystemRelease
+                # in place.  Timestamps and refs have not been updated yet, so the CodingSystemRelease
+                # represents the state of the previous import.
+                shutil.copy(self.backup_path, self.new_db_path)
+                self.backup_path.unlink()
+            else:
+                # if there was no existing db, remove the failed new db file and delete the CodingSystemRelease
+                self.cs_release.delete()
+                self.new_db_path.unlink()
+            # raise the exception
+            return False
+        # Finally, if all went well, update the coding system release fields
+        self.cs_release.import_timestamp = self.import_datetime
+        self.cs_release.import_ref = self.import_ref
+        self.cs_release.save()
+
+        if self.backup_path is not None:
+            self.backup_path.unlink()
+
+    @transaction.atomic
+    def setup_new_import_database(self):
+        print("Setting up database...")
+        # get or create the CodingSystemRelease; note we don't update an existing
+        # CodingSystemRelease with a new import_timestamp or import_ref yet, in case
+        # anything goes wrong during the import
+        # Note that this method is wrapped in transaction.atomic, but this will ONLY
+        # roll back the CodingSystemRelease object in case of errors raised in this
+        # method itself.  The __exit__ method deals with reverting the state of the
+        # release database if anything goes wrong in the import itself.
+        self.cs_release, new = CodingSystemRelease.objects.get_or_create(
```

Comment:
Under what circumstances would we be getting an already-created `CodingSystemRelease`?

Reply:
Looks like we're supporting reimporting a release.  Given the complexity this introduces (with multiple setup and teardown paths) is this worth it?

(I think the code is correct!  But I'm having a slightly hard time following it.)

Reply:
> Under what circumstances would we be getting an already-created `CodingSystemRelease`?

If we're overwriting an existing import.  It's been extremely useful in testing, but I'd hope it wouldn't be used often in production.  There could be a case where someone downloads the wrong release file and wants to re-import it.

I'll try to add some better comments!  I did make sure this is all fully covered by tests :)